File: 0981 time-based-key-value-store/solution.java
class Pair {
    int timestamp;
    String value;

    Pair(int timestamp, String value) {
        this.timestamp = timestamp;
        this.value = value;
    }
}

class TimeMap {
    Map<String,List<Pair>> map;
    public TimeMap() {
         map = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        map.putIfAbsent(key,new ArrayList<>());
        map.get(key).add(new Pair(timestamp,value));
    }
    
    public String get(String key, int timestamp) {
        if(!map.containsKey(key)) return "";
        List<Pair> list = map.get(key);
        int l=0,r=list.size()-1;
        String res = "";
        while(l<=r) {
            int mid = l+(r-l)/2;
            if(list.get(mid).timestamp <=timestamp) {
                res=list.get(mid).value;
                l=mid+1;
            } else {
                r=mid-1;
            }
        }
        return res;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */

File: 0981 time-based-key-value-store/solution.go
type TimeMap struct {
    mapp map[string][]Pair
}

func Constructor() TimeMap {
    return TimeMap{
        mapp: make(map[string][]Pair),
    }
}



func (this *TimeMap) Set(key string, value string, timestamp int)  {
    this.mapp[key] = append(this.mapp[key],Pair{Timestamp: timestamp, Value: value})
}


func (this *TimeMap) Get(key string, timestamp int) string {
    pairs := this.mapp[key]
    res := ""
    if len(pairs) !=0 {
        l,r:=0,len(pairs)-1
        for l<=r {
            mid:=l+(r-l)/2
            if pairs[mid].Timestamp <= timestamp {
                res = pairs[mid].Value
                l = mid+1
            } else {
                r = mid-1
            }
        }
    }
    return res
}





// Define a struct
type Pair struct {
    Timestamp int
    Value     string
}

// Optional constructor-like function
func NewPair(timestamp int, value string) Pair {
    return Pair{Timestamp: timestamp, Value: value}
}



/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */
